### PR TITLE
Implement master submitty DB and use it for single sign-on across courses

### LINKED ISSUE #147: "Central" DB that knows all users & all courses (renamed issue)
A single relational DB would ensure greater data consistency, and reduce the chances of data anomalies.

### PR DESCRIPTION
**Pull request by MasterOdin**
Closes #147 

This accomplishes the following things:
* Adds a central DB that contains all users added to Submitty, use this DB for handling of sessions + what users are in a course, while having a course DB
  * Use triggers to push inserts/updates from the main DB to the course DB
* Display an error page if trying to go to a course that you do not have access to (but are in the Submitty system)

Potential new issues:
* Update student feed to use the new DB structure
* Build page of courses dynamically (instead of it statically generated in setup_sample_courses.py)
  * Show all courses in system
  * Show all courses the user has access to
* Mark some courses as 'archived' so that they are no longer updated via the triggers (and marked as such on the master course list)
* Create UI for seeing all users in the system (if group is instructor+)
* Improve UI for creating user in the system (right now, if creating user in a course, no way to know he was in a different course or his details)

### CODE REVIEW

**Comment by MasterOdin** [comment]
dblink requires password when using NONSUPERUSER account. Need to discuss best way to handle this with @pbailie and @bmcutler.

**Comment by MasterOdin** [comment]
Another solution could be to set the functions as created by postgresql (who is a super user) and use the "SECURITY DEFINER" flag in function creation, then hsdbu would execute this function as superuser while not being one overall.

File: site/app/libraries/Database.php [hunk 1]
```diff
@@ -95,6 +97,14 @@ public function query($query, $parameters=array()) {
             $statement = $this->link->prepare($query);
             $statement->execute($parameters);
             $this->results = $statement->fetchAll();
+            $lower = strtolower($query);
+            if (Utils::startsWith($lower, "update") || Utils::startsWith($lower, "delete")
+                || Utils::startsWith($lower, "update")) {
```
Comment (jbarthelmess) [inline site/app/libraries/Database.php]: why are you using <code> Utils::startsWith($lower, "update") </code> twice?

Comment (MasterOdin) [inline site/app/libraries/Database.php]: Whoops, should have been insert, thanks. Fixed.

**Pull request merged**